src/Processors/Formats/Impl/Parquet/keyConditionRPNToParquetRPN.cpp

```diff
@@ -485,35 +342,16 @@ std::vector<ParquetBloomFilterCondition::ConditionElement> keyConditionRPNToParq
 
             if (found_empty_column)
             {
-                condition_elements.emplace_back(Function::ALWAYS_FALSE);
+                // todo arthur
                 continue;
             }
```

Comment:
you say that because if the set is empty, it'll be false and thus `rpn_stack.back().can_be_true` will prevent the bloom filter check??

Reply:
Yes. Maybe it would make sense to also do a redundant explicit check, to not rely on checkInRange carefully handling the kinda-special "infinite range x empty set" case. Either way it belongs in KeyCondition, I feel.

In my mind, empty set is not a special case at all, nothing would break if the normal code path runs on it (unless the code uses empty list as a special value with special meaning (e.g. "skip bloom filter check"), which I think is not a good idea in this case). Except that we may unnecessarily read the bloom filter from file, worth avoiding.

(Or KeyCondition's construction code could do the ALWAYS_FALSE thing, but I like that slightly less because (a) it means a more substantially different code path is taken depending on the data (whether the set happens to be empty, e.g. if it's a subquery), (b) it requires the set to be built (e.g. by running subquery) at KeyCondition construction time; which currently always happens anyway, but in future I can imagine deferring it to a different stage of query execution, with better progress reporting and cancellability than the early query analysis stage.)

Reply:
My mind might be tricking me, but I guess it is ok not to check for empty columns at this stage.

```
bool mayExistOnBloomFilter(const KeyCondition::BloomFilterData & condition_bloom_filter_data,
                           const KeyCondition::ColumnIndexToBloomFilter & column_index_to_column_bf)
```

will loop over the columns in the set index, and then call the overload

```
bool mayExistOnBloomFilter(const std::vector<uint64_t> & hashes, const std::unique_ptr<KeyCondition::BloomFilter> & bloom_filter)
```

which loops over the hashes, but none shall be found because empty set won't produce any hashes. If that is the case, it'll return false. If it returns false, the row group will be skipped.

That's the goal, right? Unless there is a scenario where min/max range check would return true for empty set, and then bloom filter would affect the end result.

Reply:
Two goals: (1) skip the row group (which probably already redundantly accomplished by both set_index->checkInRange and mayExistOnBloomFilter), (2) don't read the bf. For (2) I just realized checkRPNAgainstHyperrectangle's behavior is irrelevant, the empty set check would need to be in getBloomFilterFilteringColumnKeys (and it doesn't seem important; if any special effort is needed to "prepare" the set correctly in there, it's probably not worth it).